### PR TITLE
.coafile: Add python_import section

### PR DESCRIPTION
**Pull request by arjunsinghy96**
Adds python_import section to check non-builtin unused imports.
Adds ignore PyUnusedCodeBear to non-builtin imports as required
Removes unused import where necessary

Closes https://github.com/coala/coala/issues/2994

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/bearlib/abstractions/ExternalBearWrap.py [hunk 1]
```diff
@@ -6,9 +6,9 @@
 from coalib.bears.LocalBear import LocalBear
 from coala_utils.decorators import enforce_signature
 from coalib.misc.Shell import run_shell_command
-from coalib.results.Diff import Diff
 from coalib.results.Result import Result
 from coalib.results.SourceRange import SourceRange
+# Ignore PyUnusedCodeBear
```
Comment (RohanVB) [inline coalib/bearlib/abstractions/ExternalBearWrap.py]: maybe put this on the line with the import?

Comment (arjunsinghy96) [inline coalib/bearlib/abstractions/ExternalBearWrap.py]: That would make the line `85` char .. then LineLength and pep8 yell at me.

Comment (RohanVB) [inline coalib/bearlib/abstractions/ExternalBearWrap.py]: `from coalib.results.SourceRange import SourceRange # Ignore PyUnusedCodeBear` would be on line 10 right? thats 77 chars.

**Comment by NiklasMM** [comment]
What do you mean by `Adds more analysis.` in the commit message? I guess this can either be removed or be more concrete.

File: .coafile [hunk 2]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
 # linelength.
-bears = SpaceConsistencyBear, PyUnusedCodeBear, QuotesBear
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
+# PyUnusedCodeBear is run seperately beacause we need to check for unused
```
Comment (Techievena) [inline .coafile]: separately

File: .coafile [hunk 3]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
-bears = SpaceConsistencyBear, PyUnusedCodeBear, QuotesBear
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
+# PyUnusedCodeBear is run seperately beacause we need to check for unused 
+# imports. Do not set default_action to ApplyPatchAction: it may lead to some
```
Comment (Techievena) [inline .coafile]: `: it` --> `as it`

File: .coafile [hunk 4]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
+# PyUnusedCodeBear is run seperately beacause we need to check for unused 
+# imports. Do not set default_action to ApplyPatchAction: it may lead to some 
+# required imports being removed resulting in coala to behave weirdly.
```
Comment (Techievena) [inline .coafile]: `resulting in coala to behave` --> `that might result in coala behaving`

File: tests/bearlib/aspects/ClassTest.py [hunk 5]
```diff
@@ -1,5 +1,3 @@
-from coalib.bearlib.aspects import aspectclass
```
Comment (Techievena) [inline tests/bearlib/aspects/ClassTest.py]: Please don't tamper with aspectclasses, they are currently in development state and coala analysis should be done on them once they are up and running. These imports might be added keeping in mind to add more functionality later.

File: .coafile (hunk 4 above)
Comment (arjunsinghy96) [inline .coafile]: will make these changes

File: tests/bearlib/aspects/ClassTest.py (hunk 5 above)
Comment (arjunsinghy96) [inline tests/bearlib/aspects/ClassTest.py]: As these were tests and did not break anything, I removed the unused imports. Will change that if it is necessary.

**Comment by arjunsinghy96** [comment]
@adhikasp Some imports are required even if they are unused( like Languages.py only have imports). Removing them will break stuff.

**Comment by meetmangukiya** [comment]
> Removing them will break stuff.

How can an import break stuff if it is not used? Just trying to understand :)

**Comment by arjunsinghy96** [comment]
@meetmangukiya like the imports in `__init__.py` files. I removed them and lot of tests failed

File: coalib/settings/Section.py [hunk 6]
```diff
@@ -5,7 +5,7 @@
 
 from coalib.collecting.Collectors import collect_registered_bears_dirs
 from coala_utils.decorators import enforce_signature, generate_repr
-from coala_utils.string_processing import unescape
+from coala_utils.string_processing import unescape  # Ignore PyUnusedCodeBear
```
Comment (Makman2) [inline coalib/settings/Section.py]: why?

File: coalib/bearlib/abstractions/ExternalBearWrap.py (hunk 1 above)
Comment (Makman2) [inline coalib/bearlib/abstractions/ExternalBearWrap.py]: you have to explain me this ignore ;)

File: coalib/bears/Bear.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
 
 from dependency_management.requirements.PackageRequirement import (
     PackageRequirement)
+# Ignore PyUnusedCodeBear
```
Comment (arjunsinghy96) [inline coalib/bears/Bear.py]: In the doctests? :+1:

File: tests/bearlib/aspects/ClassTest.py (hunk 5 above)
Comment (Makman2) [inline tests/bearlib/aspects/ClassTest.py]: if this change works, it's good imo, one unnecessary import less :+1:

**Comment by Makman2** [comment]
Please file issues for `# Ignore PyUnusedCodeBear` to clarify on them inside comments or to avoid them if possible :+1: :)

**Comment by arjunsinghy96** [comment]
@Makman2 most of the unused imports were removable except for those in `coalib.bearlib.languages.__init__.py` and those used in `doctest`. init file ones are ignored and the later is moved to docstrings.

File: .coafile [hunk 8]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
 
+[python_imports]
+# PyUnusedCodeBear is run separately beacause we need to check for unused 
+# imports. Do not set default_action to ApplyPatchAction as it may lead to some 
+# required imports being removed that might result in coala behaving weirdly.
+
+bears = PyUnusedCodeBear
+language = Python
+
+remove_all_unused_imports = true
```
Comment (SanketDG) [inline .coafile]: maybe we want to explicitly, turn off the option "remove all unused variables" ?

or, we could just specify it in the description :+1:

Comment (arjunsinghy96) [inline .coafile]: I don't completely understand you. But(from what I understand) `remove_all_unused_imports` is off by default and we need to remove unused imports in coala. If a critical import is there we can add `Ignore` comment.

Comment (SanketDG) [inline .coafile]: Hey sorry, this fell out of my notifications, but what I mean is that `PyUnsedCodeBear` has another setting `remove_all_unused_variables` and we should account for that. Other explicitly switch it off, or just explain that it also removes unused variables.

Comment (arjunsinghy96) [inline .coafile]: @SanketDG `remove_unused_variables` is not passed as a setting to `PyUnusedCodeBear`. It is set `True` for the `autoflake` explicitly. Also, I think this issue deals with unused imports we can have an issue to have `remove_unused_variables` setting. What are your thoughts about this?

Comment (SanketDG) [inline .coafile]: I am not sure what you mean

https://github.com/coala/coala-bears/blob/master/bears/python/PyUnusedCodeBear.py#L33

^^ This clearly shows a setting

Comment (arjunsinghy96) [inline .coafile]: I meant to say that we don't pass it as a parameter to run(). We can have an issue for that.

Comment (SanketDG) [inline .coafile]: Uh you are right,  opened https://github.com/coala/coala-bears/issues/1584 :+1:

Comment (arjunsinghy96) [inline .coafile]: It would be helpful if you can give any reviews on this PR. It is pending for a long time and I think we need this feature.

**Comment by SanketDG** [comment]
Hey so the intention is to get more code analysis done in the coala repo, so I am not sure this should the issue.

Maybe we want to put `Partially fixes` or `Part of` in the commit body?

Also, rebase :)

**Comment by arjunsinghy96** [comment]
@SanketDG done. :+1:

**Comment by SanketDG** [comment]
@nemaniarjun Since this partially solves the issue, shouldn't it not close the issue automatically i.e. the closes <url> shouldn't be there? :)

**Comment by arjunsinghy96** [comment]
@SanketDG You tagged the wrong person. It will completely close #3066 . So, should I use the URL of that issue.

**Comment by SanketDG** [comment]
yes please

> You tagged the wrong person.

oops sorry :smile:

File: .coafile [hunk 9]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
 # check the line length because PEP8Bear sometimes isn't able to correct the
 # linelength.
-bears = SpaceConsistencyBear, PyUnusedCodeBear, QuotesBear
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
```
Comment (adtac) [inline .coafile]: This section could basically be called `unused_imports`.

File: .coafile [hunk 10]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
+# PyUnusedCodeBear is run separately beacause we need to check for unused
```
Comment (adtac) [inline .coafile]: typo because

File: .coafile [hunk 11]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
-bears = SpaceConsistencyBear, PyUnusedCodeBear, QuotesBear
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
+# PyUnusedCodeBear is run separately beacause we need to check for unused 
+# imports. Do not set default_action to ApplyPatchAction as it may lead to some
```
Comment (adtac) [inline .coafile]: The first line doesn't really add any new information that I can't understand from reading the lines below. Just keep the second line.

File: .coafile [hunk 12]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
...
+bears = SpaceConsistencyBear, QuotesBear
 language = Python
 preferred_quotation = '
 
 default_actions = *: ApplyPatchAction
 
+[python_imports]
+# PyUnusedCodeBear is run separately beacause we need to check for unused 
+# imports. Do not set default_action to ApplyPatchAction as it may lead to some 
+# required imports being removed that might result in coala behaving weirdly.
```
Comment (adtac) [inline .coafile]: Maybe explicitly set ApplyPatchAction to false right below this line. So that no one adds it accidently.

File: .coafile [hunk 13]
```diff
@@ -12,12 +12,22 @@ use_spaces = True
 # will be executed sequentially; also we need the LineLengthBear to double
```
Comment (adtac) [inline .coafile]: The commit message could just be `coafile: Move PyUnusedCodeBear to its own section` or something like that.

Comment (jayvdb) [inline .coafile]: That isnt particularly helpful either.

`.coafile: Activate remove_all_unused_imports`

?

btw, this placement of the commit message review will cause it to stick around. ala https://github.com/coala/coala/issues/4087
However, it does allow it to be replied to.
GitHub sucks for reviews.

File: .coafile (hunk 9 above)
Comment (jayvdb) [inline .coafile]: I would call it `flakes`, as that is closer to `autopep8` , and allows unused variables and other flakes to be included in this section.

File: .coafile (hunk 12 above)
Comment (jayvdb) [inline .coafile]: Anyone adding it accidentally will end up with a red build.

The commit message should give a quick explanation of why fixes are not being automatically applied.

File: .coafile (hunk 13 above)
Comment (adtac) [inline .coafile]: Ah I didn't realize it was inactive previously. I agree, your version sounds better.

**Comment by arjunsinghy96** [comment]
@jayvdb  @adtac I will do the following changes requested:
- Shortlog: `.coafile: Activate remove_all_unused_import`
- Rename section to `flakes`
- fix typo
- `default_actions = *:ShowPatchAction`
- Add to commit body `Automatic fixes may remove critical imports`

**Comment by jayvdb** [comment]
ack 210b08a

**Comment by jayvdb** [comment]
@rultor merge

**Pull request merged**